### PR TITLE
Lit: skill-component

### PR DESCRIPTION
**Pull request by dukio**
Merge #936 before this, there will be conflicts after the merge so tag me and I'll fix them... or if we decide not to merge Social Skill... just merge this lol

---

Converts `skill_component()` to Lit `skill-component`

Stuff changed:
- tippy will no longer be displayed if `data-tippy-content=""` (empty string), before it would still show the `*` and the tooltip but with Lit components it's much easier to handle it this way so that you can do `condition ? html'<span>tooltip</span>' : ""` without having to worry about an empty tooltip appearing
- copied `formatNumber()` from helper.js to `stats-defer.ts` (used in `skill-component.ts`)

First time using Lit so... let me know if I did anything wrong or if anything can be improved 😄

### CODE REVIEW

**Comment by nstringham** [comment]
> with Lit components it's much easier to handle it this way so that you can do `condition ? html'<span>tooltip</span>' : ""` without having to worry about an empty tooltip appearing

You should do undefined instead of empty string

**Comment by dukio** [comment]
> > with Lit components it's much easier to handle it this way so that you can do `condition ? html'<span>tooltip</span>' : ""` without having to worry about an empty tooltip appearing
> 
> You should do undefined instead of empty string

Changed in 922359dbdb0fc8be8fd4ba97010eb7b4b5d8dedb, thanks

Had to use `ifDefined()` (otherwise there was a ts error) so now the check if `data-tippy-content` is empty is no longer required, should I revert that change or can it be useful in the future?

**Comment by nstringham** [comment]
> Had to use `ifDefined()`

yep that's the recommended method for optional properties


> now the check if `data-tippy-content` is empty is no longer required, should I revert that change or can it be useful in the future?

I would probably revert it. it will be easy to add it back in if we need that in the future.

File: public/resources/ts/stats-defer.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ import { SkinViewer, createOrbitControls } from "skinview3d";
 import tippy from "tippy.js";
 
 import("./elements/inventory-view");
+import("./elements/skill-component");
```
Comment (nstringham) [inline public/resources/ts/stats-defer.ts:6]: this element is above the fold so it should not be lazily loaded

**Review by nstringham — CHANGES_REQUESTED** [review]
I left comments on a number of issues

some of them are fixed by https://github.com/dukio/SkyCrypt/pull/90 others still need to be fixed

**Comment by dukio** [comment]
> > Had to use `ifDefined()`
> 
> yep that's the recommended method for optional properties
> 
> > now the check if `data-tippy-content` is empty is no longer required, should I revert that change or can it be useful in the future?
> 
> I would probably revert it. it will be easy to add it back in if we need that in the future.

reverted in ef249fc59cde084296f964355c4bbae1c57e6eb9

File: public/resources/ts/elements/skill-component.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
+import { html, LitElement, TemplateResult } from "lit";
+import { customElement, property } from "lit/decorators.js";
+import { ifDefined } from "lit/directives/if-defined.js";
+
+import { formatNumber } from "../stats-defer";
+
+@customElement("skill-component")
+export class SkillComponent extends LitElement {
+  @property({ attribute: "skill" })
+  skill: string | undefined = undefined;
```
[the remark this answers is not in the document]
Comment (dukio) [reply public/resources/ts/elements/skill-component.ts]: ty!

File: public/resources/ts/elements/skill-component.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  >
+    ${icon.startsWith("head-")
+      ? html`<div class="item-icon custom-icon" style="background-image:url(/head/${icon.substring(5)})"></div>`
+      : html`<div class="item-icon ${icon}"></div>`}
+    ${level.level == level.maxLevel ? html`<div class="piece-shine"></div>` : undefined}
+  </div>`;
+}
+
+function skillProgressTemplate(skill: string, level: Levels) {
+  if (!("runecrafting" in calculated.levels)) {
```
[the remark this answers is not in the document]
Comment (dukio) [reply public/resources/ts/elements/skill-component.ts:89]: Progression can only be calculated if user has skills API enabled, we can check if he user has the API enabled by checking if we have runecrafting data, which is not present in the achievements (the place where we get the skills if skills API is disabled).

File: public/resources/ts/elements/skill-component.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+  let mainText = formatNumber(level.xpCurrent, true);
+  if (level.xpForNext && level.xpForNext != Infinity) {
+    mainText += ` / ${formatNumber(level.xpForNext, true)}`;
+  }
+  mainText += " XP";
+
+  return html`<div
+    class="skill-progress-text"
+    @mouseover="${(e: MouseEvent) => {
```
[the remark this answers is not in the document]
Comment (dukio) [reply public/resources/ts/elements/skill-component.ts:107]: Nope... but I see now... tomorrow I'll try to figure out how to do this

[the remark this answers is not in the document]
Comment (dukio) [reply public/resources/ts/elements/skill-component.ts]: Refactored in 1b80434

**Comment by dukio** [comment]
👋 ?

**Comment by dukio** [comment]
Converting to draft until I can resolve the conflicts (vscode detects 10.000+ changes and it's basically stuck trying to process what's going on 🙄 )

**Comment by dukio** [comment]
Resolved conflicts.

@nstringham fyi `interfact Level` was missing the property `rank: number;` ... idk if it was left out on purpose?

File: public/resources/ts/globals.d.ts [hunk 5]
```diff
@@ -157,6 +157,7 @@ interface Level {
   levelCap: number;
   uncappedLevel: number;
   levelWithProgress: number;
+  rank: number;
```
Comment (nstringham) [inline public/resources/ts/globals.d.ts:160]: should be `rank?: number;`

Comment (dukio) [reply public/resources/ts/globals.d.ts]: fixed, thanks ce5089d

Comment (nstringham) [reply public/resources/ts/globals.d.ts]: Np

**Comment by dukio** [comment]
Resolved conflict

File: public/resources/ts/elements/skill-component.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        ${level.level == level.maxLevel ? html`<div class="piece-shine"></div>` : undefined}
+      </div>
+      <div class="skill-name">
+        ${skillName} <span class="skill-level">${level.level >= 0 ? level.level : "?"}</span>
+      </div>
+      <div class="skill-bar" data-skill="${skillName}">
+        <div class="skill-progress-bar" style="--progress: ${level.level == level.levelCap ? 1 : level.progress}"></div>
+        ${"runecrafting" in calculated.levels
+          ? html`<div class="skill-progress-text">
+              ${this.progressText === "Loading..." ? this.getProgressTexts()[0] : this.progressText}
```
Comment (nstringham) [inline public/resources/ts/elements/skill-component.ts:69]: If the loading text is never displayed then there is no point in putting it there in the first place

**Review by nstringham — CHANGES_REQUESTED** [review]
Overall looks pretty good.

I have created https://github.com/dukio/SkyCrypt/pull/106 which resolves CSS merge conflicts and changes how the hover state works to something that seems more readable and performant to me

I still don't like how it takes like half a second for the skill bars to load in after everything else above the fold is loaded but that should get better as we move more of the page to client side rendering.

**Comment by dukio** [comment]
Merged Nate's dukio#106 which fixes a lot of things I had no clue how to fix... (thanks)

ready for review now, tested in local and works fine on multiple profiles I checked

**Review by MartinNemi03 — APPROVED** [review]

File: public/resources/ts/elements/skill-component.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    hoverText += " XP";
+
+    mainText = formatNumber(level.xpCurrent, true);
+    if (level.xpForNext && level.xpForNext != Infinity) {
+      mainText += ` / ${formatNumber(level.xpForNext, true)}`;
+    }
+    mainText += " XP";
+
+    return [mainText, hoverText];
```
[the remark this answers is not in the document]
Comment (nstringham) [reply public/resources/ts/elements/skill-component.ts]: fixed in https://github.com/SkyCryptWebsite/SkyCrypt/commit/0adf990b68fe86a526c500b8ab9f4348bb25f3b5

File: public/resources/ts/elements/skill-component.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+          ? html`<div class="skill-progress-text">
+              ${this.progressText === "Loading..." ? this.getProgressTexts()[0] : this.progressText}
+            </div>`
+          : undefined}
+      </div>
+    `;
+  }
+
+  private _updateProgressText(string: string) {
+    this.progressText = string;
```
[the remark this answers is not in the document]
Comment (nstringham) [reply public/resources/ts/elements/skill-component.ts]: fixed in https://github.com/SkyCryptWebsite/SkyCrypt/commit/13176fc1c312f3e0a8583798e9a9f5042e3d8a1a

File: public/resources/ts/elements/skill-component.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        ${"runecrafting" in calculated.levels
+          ? html`<div class="skill-progress-text">
+              ${this.progressText === "Loading..." ? this.getProgressTexts()[0] : this.progressText}
+            </div>`
+          : undefined}
+      </div>
+    `;
+  }
+
+  private _updateProgressText(string: string) {
```
[the remark this answers is not in the document]
Comment (nstringham) [reply public/resources/ts/elements/skill-component.ts]: fixed in https://github.com/SkyCryptWebsite/SkyCrypt/commit/13176fc1c312f3e0a8583798e9a9f5042e3d8a1a

File: public/resources/ts/elements/skill-component.ts (hunk 6 above)
Comment (nstringham) [reply public/resources/ts/elements/skill-component.ts]: fixed in https://github.com/SkyCryptWebsite/SkyCrypt/commit/13176fc1c312f3e0a8583798e9a9f5042e3d8a1a

**Review by nstringham — APPROVED** [review]
Everything looks good (except for performance as previously discussed)

**Pull request merged**